src/lib/transition_frontier_controller_tests/test_bootstrap_controller.ml

```diff
@@ -37,6 +37,88 @@ end)
+      let ancestor_prover = Ancestor.Prover.create ~max_size:max_length in
+      Thread_safe.block_on_async_exn (fun () ->
+          let%bind frontier = create_root_frontier ~max_length ~logger in
+          let genesis_root =
+            Transition_frontier.root frontier
+            |> Transition_frontier.Breadcrumb.transition_with_hash
+            |> With_hash.data
+            |> External_transition.forget_consensus_state_verification
+          in
+          let bootstrap =
+            Bootstrap_controller.For_tests.make_bootstrap ~logger
+              ~ancestor_prover ~genesis_root ~network ~max_length
+          in
+          let ledger_db =
+            Transition_frontier.For_tests.root_snarked_ledger frontier
+          in
+          let root_sync_ledger =
+            Root_sync_ledger.create ledger_db ~parent_log:logger
+          in
+          let parent_breadcrumb = Transition_frontier.best_tip frontier in
+          (* generating random breadcrumbs *)
+          let breadcrumbs_gen =
+            Quickcheck_lib.gen_imperative_list
+              ( Quickcheck.Generator.return @@ Deferred.return
+              @@ parent_breadcrumb )
+              (gen_breadcrumb ~logger)
+            |> Quickcheck.Generator.with_size ~size:num_breadcrumbs
+          in
+          let%bind breadcrumbs =
+            Deferred.all @@ Quickcheck.random_value breadcrumbs_gen
+          in
+          let input_transitions_verified =
+            List.map
+              ~f:
+                (Fn.compose With_hash.data
+                   Transition_frontier.Breadcrumb.transition_with_hash)
+              breadcrumbs
+          in
+          let envelopes =
+            List.map ~f:Envelope.Incoming.local input_transitions_verified
+          in
+          (* setting up and writing into the pipes *)
+          let reader, writer = Pipe_lib.Strict_pipe.create Synchronous in
+          let () =
+            List.iter
+              ~f:(fun x ->
+                Pipe_lib.Strict_pipe.Writer.write writer x |> don't_wait_for )
+              (List.zip_exn
+                 (List.map ~f:(fun e -> `Transition e) envelopes)
+                 (List.map
+                    ~f:(fun t -> `Time_received t)
+                    (List.init num_breadcrumbs ~f:Fn.id)))
+          in
+          (* run `sync_ledger` *)
+          let run_sync =
+            Bootstrap_controller.For_tests.sync_ledger bootstrap
+              ~root_sync_ledger ~transition_graph ~transition_reader:reader
+          in
+          let () = Pipe_lib.Strict_pipe.Writer.close writer in
+          let%bind () = run_sync in
+          (* extract saved transitions from `transition_graph` *)
+          let saved_transitions_verified =
+            Bootstrap_controller.For_tests.Transition_cache.data
+              transition_graph
+          in
+          (* compare the input transitions with saved transitions *)
+          let module Transition_set = Set.Make (External_transition.Verified) in
```

Comment:
Nit:

Can you construct the Set module for all the External_transition modules within their Make functor? You can construct the Set module by using `Comparable.Make` functor and inputting External transition into it. You can do it like the following:

```ocaml
# module Foo_and_bar : sig
    type t = { foo: Int.Set.t; bar: string }
    include Comparable.S with type t := t
  end = struct
    module T = struct
      type t = { foo: Int.Set.t; bar: string } with sexp, compare
    end
    include T
    include Comparable.Make(T)
  end;;
module Foo_and_bar :
sig
  type t = { foo : Int.Set.t; bar : string; }
  val ( >= ) : t -> t -> bool
  val ( <= ) : t -> t -> bool
  val ( = ) : t -> t -> bool

  ...

end

```

This example is from the Real World OCaml book.